// @rocket.chat/apps-engine/server/misc/Utilities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cloneDeep = require("lodash.clonedeep");
const path = require("path");
const vm = require("vm");
class Utilities {
    static deepClone(item) {
        return cloneDeep(item);
    }
    static deepFreeze(item) {
        Object.freeze(item);
        Object.getOwnPropertyNames(item).forEach((prop) => {
            // tslint:disable-next-line:max-line-length
            if (item.hasOwnProperty(prop) && item[prop] !== null && (typeof item[prop] === 'object' || typeof item[prop] === 'function') && !Object.isFrozen(item[prop])) {
                Utilities.deepFreeze(item[prop]);
            }
        });
        return item;
    }
    static deepCloneAndFreeze(item) {
        return Utilities.deepFreeze(Utilities.deepClone(item));
    }
    static transformModuleForCustomRequire(moduleName) {
        return path.normalize(moduleName).replace(/\.\.\//g, '') + '.ts';
    }
    static buildCustomRequire(files) {
        return function _requirer(mod) {
            if (files[Utilities.transformModuleForCustomRequire(mod)]) {
                const ourExport = {};
                const context = vm.createContext({
                    require: Utilities.buildCustomRequire(files),
                    exports: ourExport,
                    process: {},
                });
                vm.runInContext(files[Utilities.transformModuleForCustomRequire(mod)].compiled, context);
                return ourExport;
            }
            else {
                return require(mod);
            }
        };
    }
}
exports.Utilities = Utilities;
